rust/frontend/src/function.rs

```diff
@@ -0,0 +1,519 @@
+        if self.func.is_none() {
+            bail!("{}", ErrorKind::FunctionNotFound);
+        }
+
+        let mut ret_val = unsafe { mem::uninitialized::<ts::TVMValue>() };
+        let mut ret_type_code = 0 as c_int;
+        if self.arg_buf.is_some() {
+            let arg_buf = self.arg_buf?;
+            let mut num_args = arg_buf.len();
+            let mut values = arg_buf
+                .iter()
+                .map(|tav| tav.value.inner)
+                .collect::<Vec<ts::TVMValue>>();
+            let mut tcodes = arg_buf
+                .iter()
+                .map(|tav| tav.type_code as c_int)
+                .collect::<Vec<_>>();
+
+            if self.ret_buf.is_some() {
+                num_args = num_args + 1;
+                let ret_buf = self.ret_buf?;
+                let (ret_val, ret_type_code) = TVMRetValue::into_tvm_value(ret_buf);
+                values.append(&mut vec![ret_val.inner]);
+                tcodes.append(&mut vec![ret_type_code as c_int]);
+            }
+
+            values.truncate(num_args);
+            tcodes.truncate(num_args);
+            check_call!(ts::TVMFuncCall(
+                self.func?.handle,
+                values.as_mut_ptr(),
+                tcodes.as_mut_ptr(),
+                num_args as c_int,
+                &mut ret_val as *mut _,
+                &mut ret_type_code as *mut _
+            ));
+        } else {
+            check_call!(ts::TVMFuncCall(
+                self.func?.handle,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0 as c_int,
+                &mut ret_val as *mut _,
+                &mut ret_type_code as *mut _
+            ));
+        }
+
+        let ret = unsafe {
+            TVMRetValue::from_tvm_value(TVMValue::new(ret_val), (ret_type_code as i64).into())
+        };
+        Ok(ret)
+    }
+}
+
+/// Converts a [`Function`] to builder. Currently, this is the best way to work with
+/// TVM functions.
+impl<'a> From<Function> for Builder<'a> {
+    fn from(func: Function) -> Self {
+        Builder::new(Some(func), None, None)
+    }
+}
+
+/// Converts a mutable reference of a [`Module`] to [`Builder`].
+impl<'a: 'b, 'b> From<&'b mut Module> for Builder<'a> {
+    fn from(module: &mut Module) -> Self {
+        Builder::new(module.entry.take(), None, None)
+    }
+}
+
+unsafe extern "C" fn tvm_callback(
+    args: *mut ts::TVMValue,
+    type_codes: *mut c_int,
+    num_args: c_int,
+    ret: ts::TVMRetValueHandle,
+    fhandle: *mut c_void,
+) -> c_int {
+    let len = num_args as usize;
+    let args_list = slice::from_raw_parts_mut(args, len);
+    let type_codes_list = slice::from_raw_parts_mut(type_codes, len);
+    let mut local_args: Vec<TVMArgValue> = Vec::new();
+    // due to unsafe mem::uninitialized rustc warning about unused `value` and `tcode`.
+    let mut _value = mem::uninitialized::<ts::TVMValue>();
+    let mut _tcode = mem::uninitialized::<c_int>();
```

Comment:
underscore prefix usually means "this variable is unused," which is definitely note the case here!

Reply:
right! and that's why I wrote an explanation that without underscore, it's considered unused and most probably bc of unsafe `mem::uninitialized`, so I had to underscore it and didn't want to allow `unused_variables`

Reply:
ah, that's what the comment meant! :)

Well, in this case the linter is actually wrong, so disabling its warning the better approach. You can make the scope small, though by doing
```diff
	fhandle: *mut c_void,
) -> c_int {
+    #![allow(unused_variables)]
	 let len = num_args as usize;
```